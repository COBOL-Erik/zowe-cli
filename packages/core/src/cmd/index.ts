/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./builders/AbstractCommandBuilder";
export * from "./constants/OptionConstants";
export * from "./doc";
export * from "./handlers/FailedCommandHandler";
export * from "./help";
export * from "./profiles";
export * from "./response";
export * from "./syntax/SyntaxValidator";
export * from "./types/SecureOperationFunction";
export * from "./utils";
export * from "./yargs";

export * from "./ChainedHandlerUtils";
export * from "./CommandPreparer";
export * from "./CommandProcessor";
