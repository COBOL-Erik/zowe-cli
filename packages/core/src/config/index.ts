/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./api";
export * from "./doc";

export * from "./Config";
export * from "./ConfigAutoStore";
export * from "./ConfigBuilder";
export * from "./ConfigConstants";
export * from "./ConfigSchema";
export * from "./ProfileCredentials";
export * from "./ProfileInfo";
export * from "./ProfInfoErr";

// TODO(zFernand0): convert this to a class and deprecate old methods
// TODO(zFernand0): do a search/replace for import * as ConfigUtils
export * as ConfigUtils from "./ConfigUtils";