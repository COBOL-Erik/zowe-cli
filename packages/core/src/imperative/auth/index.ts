/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./builders/AuthCommandBuilder";
export * from "./builders/AuthLoginCommandBuilder";
export * from "./builders/AuthLogoutCommandBuilder";
export * from "./builders/CompleteAuthGroupBuilder";

export * from "./doc/IAuthHandlerApi";

export * from "./handlers/AbstractAuthHandler";
export * from "./handlers/BaseAuthHandler";
