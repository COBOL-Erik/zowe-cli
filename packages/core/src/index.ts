/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./constants";

export * from "./apiml";
export * from "./auth";
export * from "./cmd";
export * from "./console";
export * from "./config";
export * from "./error";
export * from "./expect";
export * from "./imperative";
export * from "./interfaces";
export * from "./io";
export * from "./logger";
export * from "./messages";
export * from "./operations";
export * from "./profiles";
export * from "./rest";
export * from "./security";
export * from "./settings";
export * from "./utils";
