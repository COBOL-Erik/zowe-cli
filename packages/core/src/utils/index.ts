/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/


export * from "./diff/doc/IDiffNameOptions";
export * from "./diff/doc/IDiffOptions";
export * from "./diff/doc/IWebDiffGenerator";
export * from "./diff/doc/IWebDiffManager";
export * from "./diff/DiffUtils";
export * from "./diff/WebDiffGenerator";
export * from "./diff/WebDiffManager";
export * from "./doc/IDaemonRequest";
export * from "./doc/IDaemonResponse";
export * from "./doc/ISystemInfo";
export * from "./CliUtils";
export * from "./CommonConstants";
export * from "./CoreUtils";
export * from "./DaemonRequest";
export * from "./EnvFileUtils";
export * from "./ExecUtils";
export * from "./ImperativeConfig";
export * from "./JSONUtils";
export * from "./JsUtils";
export * from "./NextVerFeatures";
export * from "./ProcessUtils";
export * from "./ProfileUtils";
export * from "./TextUtils";
