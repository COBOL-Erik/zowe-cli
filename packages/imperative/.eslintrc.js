module.exports = {
    ...require("../../.eslintrc.js"),
    "parserOptions": {
        "ecmaVersion": 12,
        "sourceType": "module",
        "project": [
            "./tsconfig.json",
            "./__tests__/tsconfig.json",
            "./tsconfig-tests.json"
        ]
    }
}