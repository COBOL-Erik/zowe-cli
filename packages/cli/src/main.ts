#!/usr/bin/env node
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IImperativeConfig, Imperative } from "@zowe/imperative";
import { Constants } from "./Constants";
import { inspect } from "util";

// TODO(Kelosky): if we remove this, imperative fails to find config in package.json & we must debug this.
const config: IImperativeConfig = {
    configurationModule: __dirname + "/imperative"
};

(async () => {
    try {
        await Imperative.init(config);
        Imperative.api.appLogger.trace("Init was successful");
        Imperative.parse();
    } catch (initErr) {
        Imperative.console.fatal("Error initializing " + Constants.DISPLAY_NAME +
            ":\n "
            + inspect(initErr));
        process.exit(1);
    }
})();
