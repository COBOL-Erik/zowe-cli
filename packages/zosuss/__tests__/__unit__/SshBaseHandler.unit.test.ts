/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IHandlerParameters, IProfile, CommandProfiles, ConnectionPropsForSessCfg } from "@zowe/imperative";
import { mockHandlerParameters } from "@zowe/cli-test-utils";
import { join, normalize } from "path";
import { Shell } from "../../src/Shell";
import { SshBaseHandler } from "../../src/SshBaseHandler";
import * as fs from "fs";

process.env.FORCE_COLOR = "0";

const UNIT_TEST_SSH_PROF_OPTS = {
    host: "somewhere.com",
    port: "22",
    user: "someone",
    password: "somesecret"
};

const UNIT_TEST_SSH_PROF_OPTS_PRIVATE_KEY = {
    host: "somewhere.com",
    port: "22",
    user: "someone",
    privateKey: normalize(join(__dirname, "..", "..", "..", "..", "..", "..", "zosuss", "__tests__", "__unit__", "__resources__", "fake_id_rsa"))
};
const UNIT_TEST_SSH_PROF_OPTS_PRIVATE_KEY_WITH_PASSPHRASE = {
    host: "somewhere.com",
    port: "22",
    user: "someone",
    privateKey: normalize(join(__dirname, "..", "..", "..", "..", "..", "..", "zosuss", "__tests__", "__unit__", "__resources__", "fake_id_rsa")),
    keyPassPhrase: "dummyPassPhrase123"
};
const UNIT_TEST_SSH_PROF_OPTS_PRIVATE_KEY_WITH_PASSPHRASE_NO_USER = {
    host: "somewhere.com",
    port: "22",
    privateKey: normalize(join(__dirname, "..", "..", "..", "..", "..", "..", "zosuss", "__tests__", "__unit__", "__resources__", "fake_id_rsa")),
    keyPassPhrase: "dummyPassPhrase123"
};


// A mocked profile map with ssh profile
const UNIT_TEST_PROFILE_MAP = new Map<string, IProfile[]>();
UNIT_TEST_PROFILE_MAP.set(
    "ssh", [{
        name: "ssh",
        type: "ssh",
        ...UNIT_TEST_SSH_PROF_OPTS
    }]
);
const UNIT_TEST_PROFILES_SSH = new CommandProfiles(UNIT_TEST_PROFILE_MAP);

const UNIT_TEST_PROFILE_MAP_PRIVATE_KEY = new Map<string, IProfile[]>();
UNIT_TEST_PROFILE_MAP_PRIVATE_KEY.set(
    "ssh", [{
        name: "ssh",
        type: "ssh",
        ...UNIT_TEST_SSH_PROF_OPTS_PRIVATE_KEY
    }]
);
const UNIT_TEST_PROFILE_MAP_PRIVATE_KEY_WITH_PASSPHRASE = new Map<string, IProfile[]>();
UNIT_TEST_PROFILE_MAP_PRIVATE_KEY_WITH_PASSPHRASE.set(
    "ssh", [{
        name: "ssh",
        type: "ssh",
        ...UNIT_TEST_SSH_PROF_OPTS_PRIVATE_KEY_WITH_PASSPHRASE
    }]
);
const UNIT_TEST_PROFILE_MAP_PRIVATE_KEY_WITH_PASSPHRASE_NO_USER = new Map<string, IProfile[]>();
UNIT_TEST_PROFILE_MAP_PRIVATE_KEY_WITH_PASSPHRASE.set(
    "ssh", [{
        name: "ssh",
        type: "ssh",
        ...UNIT_TEST_SSH_PROF_OPTS_PRIVATE_KEY_WITH_PASSPHRASE_NO_USER
    }]
);

const UNIT_TEST_PROFILES_SSH_PRIVATE_KEY = new CommandProfiles(UNIT_TEST_PROFILE_MAP_PRIVATE_KEY);
const UNIT_TEST_PROFILES_SSH_PRIVATE_KEY_WITH_PASSPHRASE = new CommandProfiles(UNIT_TEST_PROFILE_MAP_PRIVATE_KEY_WITH_PASSPHRASE);
const UNIT_TEST_PROFILES_SSH_PRIVATE_KEY_WITH_PASSPHRASE_NO_USER = new CommandProfiles(UNIT_TEST_PROFILE_MAP_PRIVATE_KEY_WITH_PASSPHRASE_NO_USER);

// Mocked parameters for the unit tests
const DEFAULT_PARAMETERS: IHandlerParameters = mockHandlerParameters({
    arguments: UNIT_TEST_SSH_PROF_OPTS,
    positionals: ["zos-uss", "issue", "ssh"],
    definition: {} as any,
    profiles: UNIT_TEST_PROFILES_SSH
});

const DEFAULT_PARAMETERS_PRIVATE_KEY: IHandlerParameters = mockHandlerParameters({
    arguments: UNIT_TEST_SSH_PROF_OPTS_PRIVATE_KEY,
    positionals: ["zos-uss", "issue", "ssh"],
    definition: {} as any,
    profiles: UNIT_TEST_PROFILES_SSH_PRIVATE_KEY
});

const DEFAULT_PARAMETERS_KEY_PASSPHRASE: IHandlerParameters = mockHandlerParameters({
    arguments: UNIT_TEST_SSH_PROF_OPTS_PRIVATE_KEY_WITH_PASSPHRASE,
    positionals: ["zos-uss", "issue", "ssh"],
    definition: {} as any,
    profiles: UNIT_TEST_PROFILES_SSH_PRIVATE_KEY_WITH_PASSPHRASE,
});
const DEFAULT_PARAMETERS_KEY_PASSPHRASE_NO_USER: IHandlerParameters = mockHandlerParameters({
    arguments: UNIT_TEST_SSH_PROF_OPTS_PRIVATE_KEY_WITH_PASSPHRASE_NO_USER,
    positionals: ["zos-uss", "issue", "ssh"],
    definition: {} as any,
    profiles: UNIT_TEST_PROFILES_SSH_PRIVATE_KEY_WITH_PASSPHRASE_NO_USER,
});

class myHandler extends SshBaseHandler {
    public async processCmd(commandParameters: IHandlerParameters): Promise<void> {
        return await Shell.executeSsh(
            this.mSession,
            commandParameters.arguments.command,
            (data: any) => commandParameters.response.console.log(Buffer.from(data))
        );
    }
}
const testOutput = "TEST OUTPUT";

describe("issue ssh handler tests", () => {

    afterEach(() => {
        jest.resetAllMocks();
    });

    it("should be able to get stdout", async () => {
        Shell.executeSsh = jest.fn(async (session, command, stdoutHandler) => {
            stdoutHandler(testOutput);
        });
        const handler = new myHandler();
        const params = Object.assign({}, ...[DEFAULT_PARAMETERS]);
        params.arguments.command = "pwd";
        await handler.process(params);
        expect(Shell.executeSsh).toHaveBeenCalledTimes(1);
        expect(testOutput).toMatchSnapshot();
    });

    it("should be able to get stdout with private key and key passphrase", async () => {
        Shell.executeSsh = jest.fn(async (session, command, stdoutHandler) => {
            stdoutHandler(testOutput);
        });
        const handler = new myHandler();
        const params = Object.assign({}, ...[DEFAULT_PARAMETERS_KEY_PASSPHRASE]);
        params.arguments.command = "echo test";
        await handler.process(params);
        expect(Shell.executeSsh).toHaveBeenCalledTimes(1);
        expect(testOutput).toMatchSnapshot();
    });
    it("should prompt user for keyPassphrase if none is stored and privateKey requires one", async () => {
        Shell.executeSsh = jest.fn(async (session, command, stdoutHandler) => {
            stdoutHandler(testOutput);
        });
        jest.spyOn(fs,"readFileSync").mockReturnValue("dummyPrivateKey");
        const handler = new myHandler();
        jest.spyOn(handler,"processCmd").mockImplementationOnce(() => {throw new Error("but no passphrase given");});
        const params = Object.assign({}, ...[DEFAULT_PARAMETERS_KEY_PASSPHRASE]);
        params.arguments.command = "echo test";
        jest.spyOn(ConnectionPropsForSessCfg as any,"getValuesBack").mockReturnValue(() => ({
            keyPassphrase: "validPassword"
        }));
        await handler.process(params);
        expect(Shell.executeSsh).toHaveBeenCalledTimes(1);
        expect(testOutput).toMatchSnapshot();
    });
    it("should reprompt user for keyPassphrase up to 3 times if stored passphrase failed", async () => {
        Shell.executeSsh = jest.fn(async (session, command, stdoutHandler) => {
            stdoutHandler(testOutput);
        });
        jest.spyOn(fs,"readFileSync").mockReturnValue("dummyPrivateKey");
        const handler = new myHandler();
        const params = Object.assign({}, ...[DEFAULT_PARAMETERS_KEY_PASSPHRASE]);
        params.arguments.command = "echo test";
        jest.spyOn(handler,"processCmd").mockImplementationOnce(() => {throw new Error("bad passphrase?");});
        jest.spyOn(ConnectionPropsForSessCfg as any,"getValuesBack").mockReturnValue(() => ({
            keyPassphrase: "validPassword"
        }));
        await handler.process(params);
        expect(Shell.executeSsh).toHaveBeenCalledTimes(1);
        expect(testOutput).toMatchSnapshot();
    });
    it("should fail if user fails to enter incorrect key passphrase in 3 attempts", async () => {
        const testOutput = "Maximum retry attempts reached. Authentication failed.";
        Shell.executeSsh = jest.fn(async (session, command, stdoutHandler) => {
            stdoutHandler(testOutput);
        });
        const handler = new myHandler();
        const params = { ...DEFAULT_PARAMETERS_KEY_PASSPHRASE };
        params.arguments.command = "echo test";
        jest.spyOn(handler, "processCmd").mockImplementation(() => {
            throw new Error("bad passphrase?");
        });
        await expect(handler.process(params)).rejects.toThrow("Maximum retry attempts reached. Authentication failed.");
        expect(handler.processCmd).toHaveBeenCalledTimes(4);
        expect(testOutput).toMatchSnapshot();
    });
    it("should prompt for user and keyPassphrase if neither is stored", async () => {
        const testOutput = "test";
        Shell.executeSsh = jest.fn(async (session, command, stdoutHandler) => {
            stdoutHandler(testOutput);
        });
        const handler = new myHandler();
        const params = { ...DEFAULT_PARAMETERS_KEY_PASSPHRASE_NO_USER };
        params.arguments.command = "echo test";
        jest.spyOn(ConnectionPropsForSessCfg as any,"getValuesBack").mockReturnValue(() => ({
            user: "someone",
            keyPassphrase: "validPassword"
        }));
        await handler.process(params);
        expect(Shell.executeSsh).toHaveBeenCalledTimes(1);
        expect(testOutput).toMatchSnapshot();
    });
    it("should be able to get stdout with privateKey", async () => {
        Shell.executeSsh = jest.fn(async (session, command, stdoutHandler) => {
            stdoutHandler(testOutput);
        });
        const handler = new myHandler();
        const params = Object.assign({}, ...[DEFAULT_PARAMETERS_PRIVATE_KEY]);
        params.arguments.command = "pwd";
        await handler.process(params);
        expect(Shell.executeSsh).toHaveBeenCalledTimes(1);
        expect(testOutput).toMatchSnapshot();
    });
    it("should be able to get stdout with privateKey", async () => {
        Shell.executeSsh = jest.fn(async (session, command, stdoutHandler) => {
            stdoutHandler(testOutput);
        });
        const handler = new myHandler();
        const params = Object.assign({}, ...[DEFAULT_PARAMETERS_PRIVATE_KEY]);
        params.arguments.command = "pwd";
        await handler.process(params);
        expect(Shell.executeSsh).toHaveBeenCalledTimes(1);
        expect(testOutput).toMatchSnapshot();
    });
});
