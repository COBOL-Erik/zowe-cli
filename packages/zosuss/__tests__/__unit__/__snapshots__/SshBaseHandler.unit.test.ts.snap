// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issue ssh handler tests should be able to get stdout 1`] = `"TEST OUTPUT"`;

exports[`issue ssh handler tests should be able to get stdout 2`] = `"TEST OUTPUT"`;

exports[`issue ssh handler tests should be able to get stdout with private key and key passphrase 1`] = `"TEST OUTPUT"`;

exports[`issue ssh handler tests should be able to get stdout with privateKey 1`] = `"TEST OUTPUT"`;

exports[`issue ssh handler tests should fail if user fails to enter incorrect key passphrase in 3 attempts 1`] = `"Maximum retry attempts reached. Authentication failed."`;

exports[`issue ssh handler tests should prompt for user and keyPassphrase if neither is stored 1`] = `"test"`;

exports[`issue ssh handler tests should prompt user for keyPassphrase if none is stored and privateKey requires one 1`] = `"TEST OUTPUT"`;

exports[`issue ssh handler tests should reprompt user for keyPassphrase up to 3 times if stored passphrase failed 1`] = `"TEST OUTPUT"`;
